smarts/env/gymnasium/utils/observation_conversion.py

```diff
@@ -0,0 +1,1050 @@
+                "heading": gym.spaces.Box(
+                    low=-math.pi, high=math.pi, shape=_WAYPOINT_SHP, dtype=np.float32
+                ),
+                "lane_index": gym.spaces.Box(
+                    low=0, high=127, shape=_WAYPOINT_SHP, dtype=np.int8
+                ),
+                "lane_offset": gym.spaces.Box(
+                    low=0, high=1e10, shape=_WAYPOINT_SHP, dtype=np.float32
+                ),
+                "lane_width": gym.spaces.Box(
+                    low=0, high=1e10, shape=_WAYPOINT_SHP, dtype=np.float32
+                ),
+                "position": gym.spaces.Box(
+                    low=-1e10, high=1e10, shape=_WAYPOINT_SHP + (3,), dtype=np.float64
+                ),
+                "speed_limit": gym.spaces.Box(
+                    low=0, high=1e10, shape=_WAYPOINT_SHP, dtype=np.float32
+                ),
+            }
+        )
+
+
+class SignalsSpaceFormat(ConfigurableSpaceFormat):
+    def __init__(self, agent_interface, count) -> None:
+        self.count = count
+        super().__init__(agent_interface)
+
+    def format(self, obs: Observation):
+        signals = obs.signals
+        des_shp = _SIGNALS_SHP
+        rcv_shp = len(signals)
+        pad_shp = max(0, des_shp[0] - rcv_shp)
+
+        if rcv_shp == 0:
+            return {
+                "state": np.zeros(des_shp, dtype=np.int8),
+                "stop_point": np.zeros(des_shp + (2,), dtype=np.float64),
+                "last_changed": np.zeros(des_shp, dtype=np.float32),
+            }
+
+        signals = [
+            (signal.state, signal.stop_point, signal.last_changed)
+            for signal in signals[: des_shp[0]]
+        ]
+        state, stop_point, last_changed = zip(*signals)
+
+        # fmt: off
+        state = np.pad(state, ((0, pad_shp)), mode='constant', constant_values=0)
+        stop_point = np.pad(state, ((0, pad_shp), (0, 0)), mode='constant', constant_values=0)
+        last_changed = np.pad(last_changed, ((0, pad_shp)), mode='constant', constant_values=0)
+        # fmt: on
+
+        return {
+            "state": state,
+            "stop_point": stop_point,
+            "last_changed": last_changed,
+        }
+
+    def active(self, agent_interface: AgentInterface) -> bool:
+        return bool(agent_interface.signals)
+
+    @property
+    def name(self):
+        return "signals"
+
+    @property
+    def space(self):
+        return gym.spaces.Dict(
+            {
+                "state": gym.spaces.Box(
+                    low=0, high=32, shape=_SIGNALS_SHP, dtype=np.int8
+                ),
+                "stop_point": gym.spaces.Box(
+                    low=-1e10, high=1e10, shape=_SIGNALS_SHP + (2,), dtype=np.float64
+                ),
+                "last_changed": gym.spaces.Box(
+                    low=0, high=1e10, shape=_SIGNALS_SHP, dtype=np.float32
+                ),
+            }
+        )
+
+
+class EnabledSpaceFormat(BaseSpaceFormat):
```

Comment:
What is the purpose of `EnabledSpaceFormat` ?

Reply:
The intention here is to determine if the agent is available in the case that the observations are returned as a boid system (i.e. `ObservationOptions.full`.)